File: algorithms/algorithms-cplusplus/leetcode/Question_0697_Degree_of_an_Array.cc
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  int findShortestSubArray(vector<int>& nums) {
    int maxf = 0;
    unordered_map<int, int> f, l, r;
    for (int i = 0; i < nums.size(); i++) {
      f[nums[i]] += 1;
      maxf = max(maxf, f[nums[i]]);
      if (l.count(nums[i]) == 0) {
        l[nums[i]] = i;
      }
      r[nums[i]] = i;
    }
    vector<int> vec;
    for (unordered_map<int, int>::iterator it = f.begin(); it != f.end(); it++) {
      if (it->second == maxf) {
        vec.push_back(it->first);
      }
    }
    int res = nums.size();
    for (int n : vec) {
      res = min(res, r[n] - l[n] + 1);
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0697_Degree_of_an_Array.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.HashMap;
import java.util.Map;

public class Question_0697_Degree_of_an_Array {
    public int findShortestSubArray(int[] nums) {
        Map<Integer, Integer> f = new HashMap<>(), l = new HashMap<>(), r = new HashMap<>();
        int maxF = 0;
        for (int i = 0; i < nums.length; i++) {
            int n = nums[i];
            f.put(n, f.getOrDefault(n, 0) + 1);
            maxF = Math.max(maxF, f.get(n));
            if (!l.containsKey(n)) {
                l.put(n, i);
            }
            r.put(n, i);
        }
        int res = nums.length;
        for (Map.Entry<Integer, Integer> e : f.entrySet()) {
            if (e.getValue() != maxF) {
                continue;
            }
            res = Math.min(res, r.get(e.getKey()) - l.get(e.getKey()) + 1);
        }
        return res;
    }
}
